proposals/nnnn-light-weight-same-type-syntax.md

```diff
@@ -0,0 +1,352 @@
+func concatenate<S : Sequence>(_ lhs: S, _ rhs: S) -> S where S.Element == String {
+  ...
+}
+```
+
+However, while `where` clauses are very general and allow complex generic requirements to be expressed, they also introduce cognitive overhead when reading and writing the declaration, and looks quite different than the concrete implementation where the type was simply written as `Array<String>`. It would be nice to have a simpler solution for cases where there is only a single same-type requirement, as above.
+
+## Proposed Solution
+
+We’d like to propose a new syntax for declaring a protocol conformance requirement together with one or more same-type requirements on the protocol's _primary associated types_. This new syntax looks like the application of a concrete generic type to a list of type arguments, allowing you to write `AsyncSequence<String>` or `AsyncSequence<[Lines]>`. This builds on the user's previous intuition and understanding of generic types and is analogous to `Array<String>` and `Array<[Lines]>`.
+
+Protocols can declare one or more primary associated types using a syntax similar to a generic parameter list of a concrete type:
+
+```swift
+protocol AsyncSequence<Element> {
+  associatedtype Iterator : AsyncIteratorProtocol
+    where Element == Iterator.Element
+  ...
+}
+
+protocol DictionaryProtocol<Key : Hashable, Value> {
+  ...
+}
+```
+
+A protocol with primary associated types can be referenced from any position where a protocol conformance requirement was previously allowed, with a list of type arguments in angle brackets. 
+
+For example, an opaque result type can now constrain the primary associated type:
+
+```swift
+func readSyntaxHighlightedLines(_ file: String) -> some AsyncSequence<[Token]> {
+  ...
+}
+```
+
+The `concatenate()` function shown earlier can now be written like this:
+
+```swift
+func concatenate<S : Sequence<String>>(_ lhs: S, _ rhs: S) -> S {
+  ...
+}
+```
+
+## Detailed design
+
+At the protocol declaration, an optional _primary associated types list_ delimited by angle brackets can follow the protocol name. When present, at least one primary associated type must be declared. Multiple primary associated types are separated by commas. Each primary associated type may optionally declare an inheritance clause. The formal grammar is amended as follows, adding an optional **primary-associated-type-list** production to **protocol-declaration**:
+
+- **protocol-declaration** → attributes<sub>opt</sub> access-level-modifier<sub>opt</sub> `protocol` protocol-name primary-associated-type-list<sub>opt</sub> type-inheritance-clause<sub>opt</sub> generic-where-clause<sub>opt</sub> protocol-body
+- **primary-associated-type-list** → `<` primary-associated-type | primary-associated-type `,` primary-associated-type-list `>`
+- **primary-associated-type** → type-name
+- **primary-associated-type** → type-name `:` type-identifier
+- **primary-associated-type** → type-name `:` protocol-composition-type
```

Comment:
Grammatically, this is identical to a `generic-parameter-list` right? Are we just duplicating this production under a different name to avoid the implication that Swift supports "generic protocols"?

Reply:
It's slightly different, because primary associated types can have a default just like ordinary associatedtype declarations (but I need to update the pitch to reflect that)